### ISSUE TITLE
Multi-tenancy - Organizations as tenants is problematic

### ISSUE THREAD

**Issue by BarryDrez** [eclipse-che/che#8623]
### Description
We are looking for a multi-tenant solution with Che.

Using Organizations as tenants solves part of the problem, but it is problematic in that there is only one key that signs all certs for all organizations due to the single realm limitations (since organizations within Che are all tied to the same realm).  For us, this is a problem as the customers typically will not accept this for security reasons.

A realm can have multiple keystores. The problem is that the keystores with their certificates and keys from all of the tenants have to go into the same realm because Che only allows one realm. Companies will not allow this. They'll want to have control over their certificates. This is a not a problem in an on premise deployment. But it is a problem in an iPaaS which services multiple companies.  

This relates to Keycloak as it is used in Che 6.  Tenants relate to different companies. For example Acme Inc would not want their keys stored along with their competitors or customers who happen to be using the same iPaaS.  In the Keycloak documentation in the section http://www.keycloak.org/docs/3.2/securing_apps/topics/oidc/java/multi-tenancy.html there is a description of how multi tenancy is achieved through multiple realms. 

Imagine an iPaaS which uses multiuser Che and the Keycloak instance that starts with Che.
There is a single Keycloak realm called Che. This Che realm is used by all tennants in this iPaaS.  A Keycloak realm can have multiple keystores each with their own public key and certificate. But a Keycloak realm can have only one active key/certificate.
Access tokens that are issued by this realm will be signed using this same certificate. 
And this certificate needs to be shared with protected resources that the clients need to access. The first problem is that the tennants will probably want their keystores isolated from other tennants.
The second problem is that tenants have no control over which key is used to sign the tokens. 

A tenant in this case is a company that is using this iPaaS. These are completely separate companies; i.e. they are not organizations within the same company.  Its not so much encryption but signature. The tokens needed to access protected resources will be signed by the realm's active keypair. When a tenant's client application needs a token to access a protected resource it is possible that the token they get will be signed using another tenant's key. 

How do we solve this problem?

**OS and version:**    
Any

**Comment by gorkem** [comment]
This is missing any information, closing

**Comment by BarryDrez** [comment]
Please reopen this issue?  Why was this closed?  There is plenty of information.

**Comment by gorkem** [comment]
Please add more information and then we can re-open. The word "Multi-tenancy" is NOT plenty of information.

**Comment by skabashnyuk** [comment]
```
For example Acme Inc would not want their keys stored along with their competitors or customers who happen to be using the same iPaaS.
```
@BarryDrez  about what exactly keys you are talking in this sentence?

**Comment by levjason** [comment]
@skabashnyuk We mean realm keys. See http://www.keycloak.org/docs/latest/server_admin/index.html#realm_keys

**Comment by skabashnyuk** [comment]
And what is the security issue to have the same realm for multiple users?

**Comment by levjason** [comment]
In a multi-tenant environment, each tenant could a different company.
Companies will not want their keys stored in the same realm as other
companies. Each tenant will want control over their keys and to keep them
separate from other tenant's keys.

On Mon, Feb 12, 2018 at 1:38 PM, Sergii Kabashniuk <notifications@github.com

**Comment by skabashnyuk** [comment]
What do you mean want to control? They should not have any control over realm private keys. That is supposed to be a secret to everybody. Only keycloak admins should have access to the private part.

Have you considered the ultimate level of isolation with individual keycloak, che-master, postgresql, infrastructure (openshift, kubernetes) copy for each organization?

**Comment by levjason** [comment]
We have considered the ultimate level of isolation. Is that Red Hat's
recommendation for building a multi-tenant iPaaS for Che?

On Mon, Feb 12, 2018 at 1:57 PM, Sergii Kabashniuk <notifications@github.com

**Comment by skabashnyuk** [comment]
Red Hat, as a company has no relation to my last statement. Just common sense tells me - if you need isolation when depending on reasons, isolate components.  Che organization is a logical level separation. Another option is infrastructure separation because of scalability or fault tolerance, etc. At this moment your motivation for me is unclear. I mean this sentence 
``` it is possible that the token they get will be signed using another tenant's key.```
I can't practically connect to the process of jwt token signing with keycloak private key and jwt token verification on che-master side with keycloak public key.
Can you rephrase your concerns without using word ```tenant```? Because it is very speculative and can have completely different meaning for different persons.

**NOTE**
If you think that your information can potentially hurt someone, please follow instruction on this page https://www.eclipse.org/security/

**Comment by levjason** [comment]
As far as I know, when JWT is using an RSA algorithm, the JWT is signed
with a prvate key. The JWT is verified on the receiving end using the
public key. The problem is that a realm supports only one active keypair
and Che only supports one realm. So the only way to have true isolation, as
you wrote, is to duplicate everything.

On Mon, Feb 12, 2018 at 2:40 PM, Sergii Kabashniuk <notifications@github.com

**Comment by skabashnyuk** [comment]
Question: why you need isolation on a level of JWT token signing? Is there any keycloak or OpenID provider document that motivates you to do that? Is there any reason why you think that this process is not secure for different users?

**Comment by levjason** [comment]
Because if there is only one realm there can be only one active keypair.
And the keypair that is suitable for one company or even organization might
not be suitable for another. Could you discuss this with your colleagues in
Keycloak development?

On Mon, Feb 12, 2018 at 3:22 PM, Sergii Kabashniuk <notifications@github.com

**Comment by skabashnyuk** [comment]
It would be better if you start this discussion with Keycloak experts because I don't understand the issue.  As far as I can see they have no recommendations in docs on this topic as well.
